csrc/mma_type.h

```diff
@@ -71,26 +81,119 @@ struct MatMulTileOptions {
+//!  latency bound kernels.
+
+#define MACRO(arch, m, n, k) \
+  arch##_##m##_##n##_##k =   \
+      MmaMacroUnderlying(MmaMacroUnderlying::Arch::arch, m, n, k)
+
+enum class MmaMacro : uint64_t {
+  NoMMA = 0,
+
+  MACRO(Turing, 16, 8, 16),
+  MACRO(Turing, 16, 16, 16),
+
+  MACRO(Ampere, 16, 8, 16),
+  MACRO(Ampere, 16, 16, 16),
+  MACRO(Ampere, 16, 8, 8),
+
+  MACRO(Hopper, 64, 8, 16),
+  MACRO(Hopper, 64, 16, 16),
+  MACRO(Hopper, 64, 24, 16),
+  MACRO(Hopper, 64, 32, 16),
+  MACRO(Hopper, 64, 40, 16),
+  MACRO(Hopper, 64, 48, 16),
+  MACRO(Hopper, 64, 56, 16),
+  MACRO(Hopper, 64, 64, 16),
+  MACRO(Hopper, 64, 72, 16),
+  MACRO(Hopper, 64, 80, 16),
+  MACRO(Hopper, 64, 88, 16),
+  MACRO(Hopper, 64, 96, 16),
+  MACRO(Hopper, 64, 104, 16),
+  MACRO(Hopper, 64, 112, 16),
+  MACRO(Hopper, 64, 120, 16),
+  MACRO(Hopper, 64, 128, 16),
+  MACRO(Hopper, 64, 136, 16),
+  MACRO(Hopper, 64, 144, 16),
+  MACRO(Hopper, 64, 152, 16),
+  MACRO(Hopper, 64, 160, 16),
+  MACRO(Hopper, 64, 168, 16),
+  MACRO(Hopper, 64, 176, 16),
+  MACRO(Hopper, 64, 184, 16),
+  MACRO(Hopper, 64, 192, 16),
+  MACRO(Hopper, 64, 200, 16),
+  MACRO(Hopper, 64, 208, 16),
+  MACRO(Hopper, 64, 216, 16),
+  MACRO(Hopper, 64, 224, 16),
+  MACRO(Hopper, 64, 232, 16),
+  MACRO(Hopper, 64, 240, 16),
+  MACRO(Hopper, 64, 248, 16),
+  MACRO(Hopper, 64, 256, 16),
+};
+
+#undef MACRO
+
+constexpr MmaMacroUnderlying::operator MmaMacro() {
+#if IS_CPP20 && !defined(__clang__)
+  // std::bit_cast for bit field is not supported by clang yet
+  return std::bit_cast<MmaMacro>(*this);
+#else
+  return static_cast<MmaMacro>(static_cast<uint64_t>(*this));
+#endif
+}
+
+constexpr MmaMacroUnderlying::MmaMacroUnderlying(MmaMacro macro) {
+#if IS_CPP20 && !defined(__clang__)
+  // std::bit_cast for bit field is not supported by clang yet
+  *this = std::bit_cast<MmaMacroUnderlying>(macro);
+#else
+  k = (uint64_t)macro & 0xFFFF; // NOLINT
+  n = ((uint64_t)macro >> 16) & 0xFFFF; // NOLINT
+  m = ((uint64_t)macro >> 32) & 0xFFFF; // NOLINT
+  arch = (Arch)((uint64_t)macro >> 48); // NOLINT
+#endif
+}
+
 //! Information for configuring and lowering mma ops
 struct MmaOptions {
-  //! Type of mma instrinsic macro to use
-  //!  This will translate to which mma intrinsic from runtime string
-  //!    to be generated to implement the mma op. The current plan
-  //!    is to have exactly one macro for each
-  //!  (arch, datatype, operand layout) triple, though there
-  //!  exists multiple possibilities for some cases, e.g. for Turing and fp16
-  //!  one can use 16_8_8 or 16_8_16.
-  //! Will consider adding more choices that the scheduler can pick from
-  //!  when our perf target becomes more fine grained, which is more likely in
-  //!  latency bound kernels.
-  enum class MacroType {
-    NoMMA = 0,
-    Ampere_16_8_16,
-    Ampere_16_16_16,
-    Turing_16_8_16,
-    Turing_16_16_16,
-    Ampere_16_8_8 // place holder for tf32
-  };
+  using MacroType = MmaMacro;
```

Comment:
I might need to do a global string replacement `MmaOptions::MacroType` -> `MmaMacro`, but this will conflict with my other PRs. So will do this clean up in a followup when my related outstanding PRs are merged.